nodejs-server/src/echoServer.ts

```diff
@@ -0,0 +1,316 @@
+    const message = request.error && request.error.message ?
+        request.error.message :
+        'Error';
+    const code = request.error && request.error.code ?
+        request.error.code :
+        grpc.status.INVALID_ARGUMENT;
+    const details =
+        request.error && request.error.details ? request.error.details : [];
+    const error = new google.rpc.Status();
+    error.code = code;
+    error.message = message;
+    error.details = details;
+    return error;
+  }
+
+  /**
+   * Helper function to build `grpc.ServiceError` out of `google.rpc.Status`.
+   */
+  private static requestToError(request:
+                                    showcaseV1Beta1.EchoRequest|
+                                showcaseV1Beta1.ExpandRequest|
+                                showcaseV1Beta1.WaitRequest):
+      grpc.ServiceError {
+    const status = EchoServer.requestToStatus(request);
+    const error = new Error(status.message) as grpc.ServiceError;
+    error.code = status.code;
+    return error;
+  }
+
+  /**
+   * Unary call example. Echoes the request.
+   */
+  echo(
+      call: grpc.ServerUnaryCall<showcaseV1Beta1.EchoRequest>,
+      callback: grpc.requestCallback<showcaseV1Beta1.EchoResponse>): void {
+    const requestId = ++this.requestCount;
+    const request = call.request;
+    this.log(requestId, 'echo request:', request);
+    if (request.content) {
+      const response = new showcaseV1Beta1.EchoResponse();
+      response.content = request.content;
+      this.log(requestId, 'echo response:', response);
+      callback(null, request);
+    } else {
+      const error = EchoServer.requestToError(request);
+      this.log(requestId, 'echo error:', error);
+      callback(error);
+    }
+  }
+
+  /**
+   * Server streaming call example. Splits the request string into a list of
+   * words returned as a stream.
+   */
+  expand(call: grpc.ServerWriteableStream<showcaseV1Beta1.ExpandRequest>):
+      void {
+    const requestId = ++this.requestCount;
+    const request = call.request;
+    this.log(requestId, 'expand request:', request);
+    this.log(requestId, 'expand writing to stream:');
+    if (!request.content) {
+      const response = new showcaseV1Beta1.EchoResponse();
+      this.log(requestId, 'expand error, response:', response);
+      call.write(response);
+      call.end();
+      return;
+    }
+    const words = request.content.split(/\s+/);
+    for (const word of words) {
+      const response = new showcaseV1Beta1.EchoResponse();
+      response.content = word;
+      this.log(requestId, 'expand response:', response);
+      call.write(response);
+    }
+    this.log(requestId, 'expand completed');
+    call.end();
+  }
+
+  /**
+   * Client streaming call example. Joins the list of words received from a
+   * stream into a string.
+   */
+  collect(
+      call: grpc.ServerReadableStream<showcaseV1Beta1.EchoRequest>,
+      callback: grpc.requestCallback<showcaseV1Beta1.EchoResponse>): void {
```

Comment:
same as other feedback, use async/await

Reply:
I think `(call, callback) => void` is the interface that gRPC expects. This `EchoServer` class is being fed to `grpc.Server.addService()` and they expect certain API.